src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs

```diff
@@ -2666,38 +2666,87 @@ internal static void SetConsoleTextAttribute(ConsoleHandle consoleHandle, WORD a
         // Return the length of a VT100 control sequence character in str starting
         // at the given offset.
         //
-        // This code only handles the most common formatting sequences, which are
-        // all of the pattern:
-        //     ESC '[' digits+ (';' digits)* 'm'
+        // This code only handles the following formatting sequences, corresponding to
+        // the patterns:
+        //     CSI params? 'm'               // SGR: Select Graphics Rendition
+        //     CSI params? '#' [{}pq]        // XTPUSHSGR ('{'), XTPOPSGR ('}'), or their aliases ('p' and 'q')
         //
-        // There are many other VT100 escape sequences, but this simple pattern
-        // is sufficient for our formatting system.  We won't handle cursor movements
-        // or other attempts at animation.
+        // Where:
+        //     CSI:    ESC '[' OR \x009b     // (C0 or C1 CSI)
+        //     params: digit+ (';' params)?
         //
-        // Note that offset is adjusted past the escape sequence.
+        // There are many other VT100 escape sequences, but these text attribute sequences
+        // (color-related, underline, etc.) are sufficient for our formatting system.  We
+        // won't handle cursor movements or other attempts at animation.
+        //
+        // Note that offset is adjusted past the escape sequence, or at least one
+        // character forward if there is no escape sequence at the specified position.
         internal static int ControlSequenceLength(string str, ref int offset)
         {
             var start = offset;
-            if (str[offset++] != (char)0x1B)
+
+            // First, check for the CSI:
+            if ((str[offset] == (char)0x1b) && (str.Length > (offset + 1)) && (str[offset + 1] == '['))
+            {
+                // C0 CSI
+                offset += 2;
+            }
+            else if (str[offset] == (char)0x9b)
+            {
+                // C1 CSI
+                offset += 1;
+            }
+            else
+            {
+                // We need to skip over at least the current character.
+                offset += 1;
                 return 0;
+            }
 
-            if (offset >= str.Length || str[offset] != '[')
+            if (offset >= str.Length)
+            {
                 return 0;
+            }
 
-            offset += 1;
-            while (offset < str.Length)
+            // Next, handle possible numeric arguments:
+            char c;
+            do
             {
-                var c = str[offset++];
-                if (c == 'm')
-                    break;
+                c = str[offset++];
```

Comment:
I'd use a temp variable in the cycle for performance instead of `offset` which is a reference.
```c#
           char c; 
           int i = offset; 
           do 
           { 
               c = str[i++];
           }
           while ((i < str.Length) && (char.IsDigit(c) || c == ';'));
           offset = i;
```

Reply:
Thank you for looking at this, @iSazonov.

You made several performance-related recommendations:

1. Use a temporary local in place of the ref int `offset` in the loop that skips control sequence parameters.
2. Put `str.Length` into a temporary local.
3. I extrapolated from (1) to use a temporary local for `offset` for a larger portion of the function.

It wasn't clear to me that these changes would improve things. I suspect the results of these changes might be small enought that they would be difficult to measure, so I haven't tried that, but I did look at the disassembly produced by the JIT compiler for an x64 Release version of the binary.

For the first case, there are more instructions and a little more stack and register pressure. Perhaps this would be made up for by eliminating a memory access; however the memory must be accessed eventually anyway, and that memory is on the stack, in the preceding frame--I don't expect it to leave L1 cache. So I don't know if it improves things, but I wouldn't guess it is much.

For the second, it looks like `str.Length` had already been put into a register, and the only thing the change does is create more stack and register pressure, so I think this could only actually make things worse (though I bet it would be difficult to measure).

For the third, it looks similar to (1), but maybe not quite so bad? But again, it doesn't look like a clear win to me.

You can see the source changes and associated disassembly [here](https://github.com/jazzdelightsme/scratch/releases/download/v1/disassembly_for_PR_10208.zip).

Reply:
@jazzdelightsme Thanks! 
The method is not call another methods so runtime can optimize the code very efficient. 
Nevertheless, we can do something. I propose to make this method local in `ControlSequenceLength` so that we get direct access to "offset" and "str" variables in `ControlSequenceLength`.